src/main/java/com/nike/cerberus/service/KmsService.java

```diff
@@ -463,4 +467,119 @@ protected boolean kmsPolicyNeedsValidation(AwsIamRoleKmsKeyRecord kmsKeyRecord,
+                }
+                return downloadPolicy(kmsCmkId, regionName, retryCount + 1);
+            }
+        }
+        return Optional.ofNullable(policy);
+    }
+
+    /**
+     * Gets all the KMS CMK ids for a given region
+     * @param regionName The region in which you want all the KMS CMK ids
+     * @return A list of of the KMS CMK ids for the requested region.
+     */
+    public Set<String> getKmsKeyIdsForRegion(String regionName) {
+        AWSKMS kms = kmsClientFactory.getClient(regionName);
+
+        Set<String> kmsKeyIdsForRegion = new HashSet<>();
+
+        String marker = null;
+        do {
+            logger.debug("Fetching keys for region: {} and marker: {}", regionName, marker);
+            ListKeysRequest listKeysRequest = new ListKeysRequest();
+            if (marker != null) {
+                listKeysRequest.withMarker(marker);
+            }
+            ListKeysResult listKeysResult = kms.listKeys(listKeysRequest);
+            listKeysResult.getKeys().forEach(keyListEntry -> kmsKeyIdsForRegion.add(keyListEntry.getKeyId()));
+            marker = listKeysResult.getNextMarker();
+        } while (marker != null);
+
+        return kmsKeyIdsForRegion;
+    }
+
+    /**
+     * Downloads the KMS CMK policies to determine if the key was created by this.
+     *
+     * @param allKmsCmkIdsForRegion The list of all the KMS Key.
+     * @param regionName The region.
+     * @return The list of KMS Keys that were created by this.
+     */
+    public Set<String> filterKeysCreatedByKmsService(Set<String> allKmsCmkIdsForRegion, String regionName) {
+        Set<String> keysCreatedByThis = new HashSet<>();
+        int numberOfKeys = allKmsCmkIdsForRegion.size();
+        String[] kmsCmkIds = allKmsCmkIdsForRegion.toArray(new String[numberOfKeys]);
+
+        logger.info("Preparing to download and analyze {} key policies in region: {} to " +
+                "determine what keys where created by this env", numberOfKeys, regionName);
+
+        for (int i = 0; i < numberOfKeys; i++) {
+            logger.debug("Processed {}% of keys for region: {}", Math.ceil((double) i / numberOfKeys * 100), regionName);
+            String kmsCmkId = kmsCmkIds[i];
+            if (wasKeyCreatedByKmsService(kmsCmkId,regionName)) {
+                keysCreatedByThis.add(kmsCmkId);
+            }
+        }
+
+        return keysCreatedByThis;
+    }
+
+    /**
+     * Checks the KMS CMK policy to see if it was created by this environments cms cluster.
+     * @param kmsCmkId The KMS CMK id.
+     * @param regionName The region.
+     * @return true if the kms key was created by CMS for the current env
+     */
+    private boolean wasKeyCreatedByKmsService(String kmsCmkId, String regionName) {
+        boolean wasCreatedByThis = false;
+
+        logger.debug("Downloading policy for key: {} in region: {}", kmsCmkId, regionName);
+
+        Optional<Policy> policyOptional = downloadPolicy(kmsCmkId, regionName, 0);
```

Comment:
Are there any rate limit concerns in calling `GetKeyPolicy` n times for n keys every time this job is run?

Reply:
Yes, which is why in the downloadPolicy method, I have retries with exp back off, it retires 12 times and backs off and eventually returns an empty optional if it still fails.

The side effect of missing a key, is that the orphaned key wouldn't get cleaned up till the next job run.

Reply:
When, I ran this during my user acceptance testing, it never got throttled or blocked from rate limiting.

get key policy has a limit of 30 per sec, and it has a little latency, so in practice with this being a single thread, I have not seen that limit being reached. How ever if this was run by more than one env at once it could hit that limit. but should be fine because of the retries with back off.

Reply:
I was more wondering what happens to Cerberus while this job is running.

If the job is running and Cerberus needs to call `GetKeyPolicy` will Cerberus be rate limited? Will Cerberus also fail gracefully / retry like this method does?

Reply:
Also, you bring up a good point about multiple environments running at the same time.

Maybe that's unlikely since the runtime of the job is based on deploy time of the EC2 instances?